### PR TITLE
fix: allow upgrading when history has nothing but failures

### PR DESCRIPTION
**Pull request by technosophos**
**This is a Work In Progress. DO NOT MERGE IT.**

It has long been possible for people to get their releases into a state where they could no longer successfully upgrade because they had failed enough times that no deployed releases are left in their history.

This can happen in the following ways:

- They can install and fail (without `--atomic`), which leaves them with only one failed release
- They can install successfully, but have supplied `--wait`, and failed when a pod did not come up (which will mark the deployment failed)
- They can set their history to some low-ish number, and then perform enough failed releases in a row that no `deployed` ones exist

And there are probably others. All of these cases are exacerbated by CI/CD systems that naively install and then don't clean up after failures.

See #5595 for context

## The Approach

The approach I took here was to attempt to leverage Helm 3's 3-way diff to recover from a broken release. (Note: This approach is not possible in Helm 2.)

In the usual `prepareUpdate`, Helm attempts to load the last _deployed_ release from history. If none is found, the deployment fails. The proposed fix here extends that logic:

- If the deployed release exists, use that release
- If a deployed release does NOT exist, try to load the last release and use that, regardless of whether it failed
- If both loads fail, the release fails

The wager here is that even if there is a failed release, the 3-way diff will be able to detect the changes in cluster and recover from the potentially inaccurate release manifest.

I am proposing this as a bug fix, as it fixes a case that cannot currently be accounted for at all (when history gets truncated, and only failed releases are left). It seems to me that if this route does not introduce any anomalous behavior, it is right to make it the default action taken. I think it would be a patently bad idea to add yet another flag to the already misleading jumble of flags hanging off of `helm update`.

_We need to look very carefully at this approach._
* Does it solve the problem of history + failed releases?
* Does it break expected behavior?
* Does it introduce new possible bugs by diffing against a failed manifest?

## A few misc notes:

- This should work fine paired with `--atomic`, since it will not trigger the atomic rollback condition unless there is a new failure.
- This should work with `--cleanup-on-fail`, which will remove broken resources, since this continues to use the existing diff logic
- This will work with any version of --history-max greater than or equal to 1.
- When `--reuse-values` is set, this will use values that were set on the last failed update
- How do we verify its behavior?

## Example

I have a chart called `alwaysfails`, which is basically `helm create alwaysfails` with the `values.yaml` setting `image` to `nosuchimage9999`.

```console
# Run this command as often as you like
$ helm3 upgrade --install alwaysfails ./alwaysfails --debug --wait --timeout=5s
# Then remove the `--wait --timeout=5s`
$ helm3 upgrade --install alwaysfails ./alwaysfails --debug
```
I ran three failed releases, and a successful 4th.

Since removing the `wait` will mark the deployment successful, you will see this in history:

```console
$ helm3 history alwaysfails
REVISION	UPDATED                 	STATUS    	CHART            	APP VERSION	DESCRIPTION
1       	Mon Apr 13 19:33:24 2020	failed    	alwaysfails-0.1.0	1.16.0     	Release "alwaysfails" failed: timed out waiting for the condition
2       	Mon Apr 13 19:33:32 2020	failed    	alwaysfails-0.1.0	1.16.0     	Upgrade "alwaysfails" failed: timed out waiting for the condition
3       	Mon Apr 13 19:33:39 2020	superseded	alwaysfails-0.1.0	1.16.0     	Upgrade "alwaysfails" failed: timed out waiting for the condition
4       	Mon Apr 13 19:35:27 2020	deployed  	alwaysfails-0.1.0	1.16.0     	Upgrade complete
```

(Note that it does currently mark the recovered release as `superseded` instead of `failed`. We could change that. Not sure what the utility is either way.)

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

### CODE REVIEW

**Comment by marckhouzam** [comment]
I just learned about Draft PRs today, if you're interested: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

File: pkg/storage/storage.go [hunk 1]
```diff
@@ -107,6 +107,14 @@ func (s *Storage) ListDeployed() ([]*rspb.Release, error) {
 	})
 }
 
+// NoDeployedReleases describes the situation where none of the releases in the history are deployed.
+//
+// Aggressive pruning can mask the fact that there is a deployed release, but that it
+// was cleaned up by the history pruner.
+type NoDeployedReleases string
+
+func (e NoDeployedReleases) Error() string { return string(e) }
```
Comment (jlegrone) [inline pkg/storage/storage.go:116]: Assuming we're ok with targeting go 1.13, we could use wrapped errors: https://blog.golang.org/go1.13-errors

https://play.golang.org/p/4ryS6CCiPCa

**Comment by liuming-dev** [comment]
similar PR #7767 about rollback, need to get reviewer

**Comment by bacongobbler** [comment]
With Helm 2, @adamreese and I paired on this particular case. It's a doozy.

Together, we introduced https://github.com/helm/helm/pull/3097 to prevent cases where a failed deployment could not be reconciled simply because Helm 2 could not reconcile against the live state of the cluster.

This is likely the most controversial pull request we ever merged into Helm as seen with https://github.com/helm/helm/issues/3353#issuecomment-358367529. For posterity, those issues were #2257, #2341, #2346, #2437, and #3068, but there are likely others.

Now that we can reconcile against the live state of the cluster in Helm 3, I think it's fine to relax this constraint as we can at least perform _some form_ of reconciliation with the last failed release.

If you wouldn't mind looking through those past issues and see if there may be any cases that might not be addressed by this proposal, that'd be wonderful. I'm all for this but perhaps there may be some case I'm forgetting back when Helm 2 used to diff against the last failed release that might be worth looking into.

**Comment by yinzara** [comment]
This is solved by #7653 as well.

**Comment by technosophos** [comment]
Yeah, it looks like this takes a pretty similar approach to #7653, but without introducing a new flag. When I looked at that PR, I got about as far as the help text, and the description made it sound  highly dangerous. The help text in that PR suggest that that version is not production-safe... and I am wondering why @yinzara.

I'm getting concerned that the sheer number of combinatorics on the `helm upgrade` command are making it so complicated that we are never going to be able to solve this problem. (And, let's be honest, a huge part of this problem is just the design of Kubernetes.)

For example, depending on which combination of `--install` `--atomic`, `--wait`, `--force`, `--reuse-values`, `--recreate`, `--timeout`, you can get the exact same chart + values to end up in any number of different states over time. We somehow have to get the default Helm behavior to be sane enough that this is not the responsibility of the user to untangle all of the flags.

Issue #5595 is a good example: People naively think that because they all got a similar error message, they are all hitting the same "bug." But if you read through the thread, you will see there there are dozens of minuscule differences in flags + history depth + expectations that are leading to a variety of actual runtime conditions that are "not desired". In the end, releases they believe should be running or recoverable are not -- but the causes are variegated, massively compounded by the complexity of all of the flags.

I don't particularly have strong feelings about whether it is better to go ahead with a slightly modified version of #7653 or with this. But I am going to spend some time analyzing how dangerous this approach actually is -- and whether the new diffing strategy in Helm 3 can actually do an adequate job. If it can't then perhaps we do use #7653 as-is and just let people live on the edge. (Which, no doubt, will just produce new acrimonious threads with people accusing us of neglect, ignorance, and malice like they are in #5595.) But if we can show this solution to be reasonably effective, I think we should fix it as part of the core.

Finally, I am still on the fence about `--atomic`. Maybe we have to start telling people, "look, there is _no recovery story_ for failing a `helm install` unless you use `--atomic`." That actually sounds reasonable to me. But I have reservations that `--atomic` is all that reliable in all cases. Rollbacks are not time machines (as @bridgetkromhout likes to say). They come with their own set of difficulties. Are they acceptable difficulties? Perhaps. Might we be able to just skip this particular problem and rely on `--atomic`? I am not sure.

**Comment by technosophos** [comment]
The solution in #7767 is to just treat the current release as the last release. That is an interesting proposal as well, though I think that in some cases it will orphan objects that were created by a previous release. Might be good to investigate that approach as well, though.

File: pkg/action/upgrade.go [hunk 2]
```diff
@@ -196,6 +213,7 @@ func (u *Upgrade) prepareUpgrade(name string, chart *chart.Chart, vals map[strin
 	if len(notesTxt) > 0 {
 		upgradedRelease.Info.Notes = notesTxt
 	}
+	u.cfg.Log("Manifest: %s", manifestDoc)
```
Comment (adamreese) [inline pkg/action/upgrade.go:216]: Heavy log entry

Comment (technosophos) [reply pkg/action/upgrade.go:216]: I will remove.

**Comment by technosophos** [comment]
Okay @adamreese and I paired on this for a few hours today, and looked deeply into the patching code. We did find a few minor things that needed fixing. Specifically, the diffing logic was too likely to leave the release in an unstable state because it was too conservative in deletions. So we are looking into a fix for that, but will do that in a separate PR.

We also looked at the case where the new chart added an object that the old (possibly failed) release did not have, but that exists in the cluster. We decided to retain the current behavior, which is that this case _should_ fail. Otherwise we could accidentally adopt a resource that Helm was not intended to manage. But this could present an edge case if we go ahead with this proposal.

@yinzara are you interested in updating your PR to remove the flag? And maybe to use a typed error? After pair-reviewing this with @adamreese, we think this approach will work well enough that we don't need a flag. And since your patch landed before mine, I'm good with continuing with yours. We left a couple of comments on your PR, but they were all easy to fix -- especially if the u.Replace flag is removed.

We think only the following release statuses are acceptable to recover from (so both of our proposed fixes are wrong as-is):

- Failed
- Superseded

But it is unclear whether you can actually ever get a `Superseded` without first finding a `Deployed`. So probably _just_ `Failed`. (`Deleted` and all of the `Pending*` ones will introduce additional issues in the diff/merge logic.)

**Comment by yinzara** [comment]
I'm happy to update the PR to whatever you'd like.  I'm confused what you mean by using a typed error instead of a flag?

**Comment by yinzara** [comment]
Oh wait. I think you're saying to just make this the default behavior and to not introduce the flag but limit it only to the Failed and Superseded statuses.  No issues. I'll handle that.  And what you mean a typed error? I'm not 100% sure what you're referring to. Which error? You mean if there is only one in another status? or?

**Comment by technosophos** [comment]
Closing in favor of #7653

**Pull request closed without merging**